File: haskell/nbe-hoas.hs
{-# LANGUAGE BangPatterns #-}

module Main where

data Term
  = Var !Int
  | Lam !Int !Term
  | App !Term !Term
  | Let !Int !Term !Term

data Value
  = VVar !Int
  | VApp !Value !Value
  | VLam !Int (Value -> Value)

data Env
  = Cons !Int !Value !Env
  | Nil

data Names
  = NameCons !Int !Names
  | NameNil

maxOfNames :: Names -> Int -> Int
maxOfNames names !acc =
  case names of
    NameCons ident rest -> maxOfNames rest (if acc > ident then acc else ident)
    NameNil -> acc

namesOfEnv :: Env -> Names -> Names
namesOfEnv env acc =
  case env of
    Cons ident _ env1 -> namesOfEnv env1 (NameCons ident acc)
    Nil -> acc

fresh :: Names -> Int
fresh n = maxOfNames n 0 + 1

lookupEnv :: Env -> Int -> Value
lookupEnv env ident =
  case env of
    Cons ident1 val env1 ->
      if ident == ident1
        then val
        else lookupEnv env1 ident
    Nil -> VVar ident

eval :: Env -> Term -> Value
eval env term =
  case term of
    Var ident -> lookupEnv env ident
    App t u -> vapp (eval env t) (eval env u)
    Lam xId t ->
      VLam xId (\u -> eval (Cons xId u env) t)
    Let xId t u ->
      let v = eval env t
       in eval (Cons xId v env) u

vapp :: Value -> Value -> Value
vapp v1 v2 =
  case v1 of
    VLam _ body -> body v2
    _ -> VApp v1 v2

quoteValue :: Names -> Value -> Term
quoteValue n v =
  case v of
    VVar ident -> Var ident
    VApp v1 v2 -> App (quoteValue n v1) (quoteValue n v2)
    VLam _ f ->
      let y = fresh n
          n2 = NameCons y n
       in Lam y (quoteValue n2 (f (VVar y)))

normForm :: Env -> Term -> Term
normForm env t =
  let names = namesOfEnv env NameNil
      v = eval env t
   in quoteValue names v

five :: Term
five =
  Lam
    0
    ( Lam
        1
        ( App
            (Var 0)
            ( App
                (Var 0)
                ( App
                    (Var 0)
                    (App (Var 0) (App (Var 0) (Var 1)))
                )
            )
        )
    )

add :: Term
add =
  Lam
    0
    ( Lam
        1
        ( Lam
            2
            ( Lam
                3
                ( App
                    (App (Var 0) (Var 2))
                    (App (App (Var 1) (Var 2)) (Var 3))
                )
            )
        )
    )

mul :: Term
mul =
  Lam
    0
    ( Lam
        1
        ( Lam
            2
            ( Lam
                3
                ( App
                    (App (Var 0) (App (Var 1) (Var 2)))
                    (Var 3)
                )
            )
        )
    )

ten :: Term
ten = App (App add five) five

twenty :: Term
twenty = App (App add ten) ten

hundred :: Term
hundred = App (App mul ten) ten

twoHundred :: Term
twoHundred = App (App add hundred) hundred

thousand :: Term
thousand = App (App mul hundred) ten

term200000 :: Term
term200000 = App (App mul twoHundred) thousand

term4000000 :: Term
term4000000 = App (App mul twenty) term200000

nfToInt :: Term -> Int -> Int
nfToInt t !acc =
  case t of
    Lam _ x -> nfToInt x acc
    App _ x -> nfToInt x (acc + 1)
    _ -> acc

nbeTest :: Int
nbeTest =
  let n = normForm Nil term4000000
   in nfToInt n 0

main :: IO ()
main = do
  let n = nbeTest
  if n /= 4000000
    then error ("FAIL: expected 4000000, got " ++ show n)
    else pure ()

File: ocaml/nbe-hoas.ml
type term =
  | Var of int
  | Lam of int * term
  | App of term * term
  | Let of int * term * term

type value =
  | VVar of int
  | VApp of value * value
  | VLam of int * (value -> value)

type env = Cons of int * value * env | Nil

let rec max_of_names names acc =
  match names with
  | id :: rest -> max_of_names rest (max acc id)
  | [] -> acc

let rec names_of_env env acc =
  match env with Cons (id, _, rest) -> names_of_env rest (id :: acc) | Nil -> acc

let fresh names = max_of_names names 0 + 1

let rec lookup env id =
  match env with
  | Cons (id1, value, rest) -> if id = id1 then value else lookup rest id
  | Nil -> VVar id

let rec eval env = function
  | Var id -> lookup env id
  | App (t, u) -> vapp (eval env t) (eval env u)
  | Lam (id, body) -> VLam (id, fun value -> eval (Cons (id, value, env)) body)
  | Let (id, t, u) -> eval (Cons (id, eval env t, env)) u

and vapp v1 v2 = match v1 with VLam (_, body) -> body v2 | _ -> VApp (v1, v2)

let rec quote names = function
  | VVar id -> Var id
  | VApp (v1, v2) -> App (quote names v1, quote names v2)
  | VLam (_, body) ->
      let y = fresh names in
      Lam (y, quote (y :: names) (body (VVar y)))

let norm_form env term = quote (names_of_env env []) (eval env term)

let lam id body = Lam (id, body)
let app f x = App (f, x)

let five =
  lam 0 (lam 1 (app (Var 0) (app (Var 0) (app (Var 0) (app (Var 0) (app (Var 0) (Var 1)))))))

let add =
  lam 0 (lam 1 (lam 2 (lam 3 (app (app (Var 0) (Var 2)) (app (app (Var 1) (Var 2)) (Var 3))))))

let mul =
  lam 0 (lam 1 (lam 2 (lam 3 (app (app (Var 0) (app (Var 1) (Var 2))) (Var 3)))))

let ten = App (App (add, five), five)
let twenty = App (App (add, ten), ten)
let hundred = App (App (mul, ten), ten)
let two_hundred = App (App (add, hundred), hundred)
let thousand = App (App (mul, hundred), ten)
let term_200000 = App (App (mul, two_hundred), thousand)
let term_1000000 = App (App (mul, five), term_200000)

let rec nf_to_int term acc =
  match term with
  | Lam (_, body) -> nf_to_int body acc
  | App (_, body) -> nf_to_int body (acc + 1)
  | _ -> acc

let nbe_test seed = nf_to_int (norm_form Nil term_1000000) seed

let () =
  let rec rounds i acc = if i = 6 then acc else rounds (i + 1) (acc + nbe_test i) in
  let result = rounds 0 0 in
  if result <> 6_000_015 then (
    Printf.eprintf "FAIL: expected 6000015, got %d\n" result;
    exit 1)
